### ISSUE TITLE
[Bug]: Stuck on "Downloading driver v545.23.08" on TrueNAS-SCALE-24.04.1.1

### ISSUE THREAD

**Issue by WillKirkmanM** [Steam-Headless/docker-steam-headless#152]
### Describe the Bug

Running the container from [Truecharts](https://truecharts.org/) whether you have host networking or not keeps 

### Screenshots

![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/842934f3-521a-45a5-b67d-e61a272e3dba)

### Version

.

### Platform

.

### Relevant log output

```Shell
2024-07-01 21:49:01.741848+01:00Build: [2024-05-25 02:40:32] [master] [47f6f7a176ee0c6f6c870c29397cc1a8d6d57839] [debian]
2024-07-01 21:49:01.742001+01:002024-07-01T21:49:01.742001652+01:00
2024-07-01 21:49:01.742016+01:00[34m[ /etc/cont-init.d/10-setup_user.sh: executing... ][0m
2024-07-01 21:49:01.742931+01:00[35m**** Configure default user ****[0m
2024-07-01 21:49:01.742939+01:00[36m  - Setting default user uid=568(default) gid=568(default)[0m
2024-07-01 21:49:02.800794+01:00[36m  - Adding default user to any additional required device groups[0m
2024-07-01 21:49:02.801519+01:00[36m  - Adding user 'default' to group: 'video'[0m
2024-07-01 21:49:03.234809+01:00[36m  - Adding user 'default' to group: 'audio'[0m
2024-07-01 21:49:03.269638+01:00[36m  - Adding user 'default' to group: 'input'[0m
2024-07-01 21:49:03.293101+01:00[36m  - Adding user 'default' to group: 'pulse'[0m
2024-07-01 21:49:03.312581+01:00[36m  - Adding user 'default' to group: 'render' for device: /dev/input/event0[0m
2024-07-01 21:49:03.356536+01:00[36m  - Adding user 'default' to group: 'video' for device: /dev/dri/card0[0m
2024-07-01 21:49:03.368261+01:00[36m  - Adding user 'default' to group: 'avahi' for device: /dev/dri/renderD128[0m
2024-07-01 21:49:03.385212+01:00[36m  - Setting umask to 0022[0m
2024-07-01 21:49:03.385234+01:00[36m  - Create the user XDG_RUNTIME_DIR path '/tmp/.X11-unix/run'[0m
2024-07-01 21:49:03.589036+01:00[36m  - Setting ownership of all log files in '/home/default/.cache/log'[0m
2024-07-01 21:49:03.590425+01:00[36m  - Setting root password[0m
2024-07-01 21:49:03.601483+01:00[36m  - Setting user password[0m
2024-07-01 21:49:03.615646+01:00[34mDONE[0m
2024-07-01 21:49:03.615665+01:002024-07-01T21:49:03.615665430+01:00
2024-07-01 21:49:03.615670+01:00[34m[ /etc/cont-init.d/11-setup_sysctl_values.sh: executing... ][0m
2024-07-01 21:49:03.616419+01:00[35m**** Configure some system kernel parameters ****[0m
2024-07-01 21:49:03.616948+01:00[36m  - The vm.max_map_count is already greater than '524288'[0m
2024-07-01 21:49:03.616958+01:00[34mDONE[0m
2024-07-01 21:49:03.616965+01:002024-07-01T21:49:03.616965174+01:00
2024-07-01 21:49:03.616969+01:00[34m[ /etc/cont-init.d/30-configure_dbus.sh: executing... ][0m
2024-07-01 21:49:03.617712+01:00[35m**** Configure container dbus ****[0m
2024-07-01 21:49:03.618032+01:00[36m  - Container configured to run its own dbus[0m
2024-07-01 21:49:03.816487+01:00[34mDONE[0m
2024-07-01 21:49:03.816504+01:002024-07-01T21:49:03.816504098+01:00
2024-07-01 21:49:03.816508+01:00[34m[ /etc/cont-init.d/30-configure_udev.sh: executing... ][0m
2024-07-01 21:49:03.817257+01:00[35m**** Configure udevd ****[0m
2024-07-01 21:49:03.817272+01:00[36m  - Disable udevd - /sys is mounted RO[0m
2024-07-01 21:49:03.818154+01:00[36m  - Enable dumb-udev service[0m
2024-07-01 21:49:03.819692+01:00[36m  - Ensure the default user has permission to r/w on input devices[0m
2024-07-01 21:49:03.820111+01:00[34mDONE[0m
2024-07-01 21:49:03.820121+01:002024-07-01T21:49:03.820121963+01:00
2024-07-01 21:49:03.820126+01:00[34m[ /etc/cont-init.d/40-setup_locale.sh: executing... ][0m
2024-07-01 21:49:03.820837+01:00[35m**** Configure local ****[0m
2024-07-01 21:49:03.822247+01:00[36m  - Locales already set correctly to en_US.UTF-8 UTF-8[0m
2024-07-01 21:49:03.822262+01:00[34mDONE[0m
2024-07-01 21:49:03.822266+01:002024-07-01T21:49:03.822266535+01:00
2024-07-01 21:49:03.822269+01:00[34m[ /etc/cont-init.d/50-configure_pulseaudio.sh: executing... ][0m
2024-07-01 21:49:03.822918+01:00[35m**** Configure pulseaudio ****[0m
2024-07-01 21:49:03.822982+01:00[36m  - Enable pulseaudio service.[0m
2024-07-01 21:49:03.824053+01:00[36m  - Configure pulseaudio to pipe audio to a socket[0m
2024-07-01 21:49:04.391087+01:00[34mDONE[0m
2024-07-01 21:49:04.391106+01:002024-07-01T21:49:04.391106632+01:00
2024-07-01 21:49:04.391111+01:00[34m[ /etc/cont-init.d/60-configure_gpu_driver.sh: executing... ][0m
2024-07-01 21:49:04.471189+01:00[35m**** Found Intel device 'Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz To Be Filled By O.E.M. CPU @ 2.8GHz' ****[0m
2024-07-01 21:49:04.471283+01:00[36m  - Mesa has already been installed into this container[0m
2024-07-01 21:49:04.471328+01:00[35m**** No AMD device found ****[0m
2024-07-01 21:49:04.471381+01:00[35m**** Found NVIDIA device 'NVIDIA GeForce RTX 2060' ****[0m
2024-07-01 21:49:04.474021+01:00[36m  - Downloading driver v545.23.08[0m
```
```

**Comment by alansari** [comment]
Steps to solve this issue taken from another github issue: 
1: wget https://developer.download.nvidia.com/compute/cuda/12.3.2/local_installers/cuda_12.3.2_545.23.08_linux.run
2: extract the aformentioned run file to get the driver installer's .run with ./cuda*.run --tmpdir=. --extract=ext-fldr flag
3: Copy from the ext-fldr and renamed the .run file in the last step to the dataset used for /home/default so the full path within the container would be /home/default/Downloads/NVIDIA_545.23.08.run
4: restart the container and Profit!

**Comment by WillKirkmanM** [comment]
Thanks for your solution! There are some things I would like to clarify

# On Truenas Scale

Previously I could not enter the filesystem as they were PVC so I had to map it into an SMB share:
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/f85c3add-d1d5-46de-8dd4-ab84064fb147)
If I access the terminal, I get this:
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/6c73d1c0-ea3c-4c21-b9b6-447a19b4d972)

Then tried it on the TrueNas Scale shell which resulted in **ERROR: extract: path must be absolute**. If I get around the path it says that GCC was not found which is unable to install because Truenas has disabled it 🙄 
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/186d63f5-bbdd-4153-8f2c-fc94cea7338b)

![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/5450a821-a78d-4f52-a6c1-ecc38cfcfbd5)

# Arch Linux WSL Virtual Machine
If I try it on a virtual machine on Windows using WSL, I get a different error
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/f7840e90-5b52-4d52-9e9e-fb21a0ff0260)

Once again, thank you for voluntarily trying to fix my cluelessness!

**Comment by alansari** [comment]
Actually the recommended method of installing on truenas is jailmaker with host networking, a guide can be found in the steam-headless discord pinned threads.  Only reason i think its not on here is that the next version of truenas is ditching k3s and migrating to docker compose with electric eel.

**Comment by WillKirkmanM** [comment]
# Solution
> [!NOTE]
> TL;DR: https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/releases/tag/cuda

@alansari is completely correct that it is recommended to install `steam-headless` with [jailmaker](https://github.com/Jip-Hop/jailmaker) and pass throw the NIC with host networking. That guide can be found [on this Discord message](https://discord.com/channels/819327740279914516/930187988925837402/1245658927262339125) as a thread. 

## Why is it stuck on "Downloading Driver..."?
The problem has also been covered in this GitHub issue https://github.com/Steam-Headless/docker-steam-headless/issues/148 where @alansari states (https://github.com/Steam-Headless/docker-steam-headless/issues/148#issuecomment-2151427723):
> v545.23.08 is not hosted on [download.nvidia.com/XFree86/Linux-x86_64](https://download.nvidia.com/XFree86/Linux-x86_64) though thats what is offered on truenas. You wil need to get the run file and rename it to NVIDIA_545.23.08.run and place it in your dataset you mounted for /home/default/Downloads/. Or you could downgrade truenas to a previous snapshot/version from the grub boot manager. Or jump the gun to electric eel and migrate to useing native docker compose to spin up the container.
edit: fingers crossed today's schedualed maitenence release includes a diff nvidia driver release

Going to the NVIDIA repository shows everything but the needed `v545.23.08` version
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/59221e09-8b31-42cc-9c35-b3043ca9d8aa).

If you attempted to download an older / newer driver it would show a `failed to initialise nvml driver/library version mismatch` error when using `nvidia-smi` and the [forums didn't seem much help](https://forums.developer.nvidia.com/t/545-23-08-is-not-listed-in-the-nvidia-driver-downloads-page/273181).

## Saving Grace?
It turns out that the dump of the NVIDIA driver `545.23.08` was cutely stored in a [GitHub repository from flathub](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/releases/tag/cuda) where you can [download the 545.23.08 driver's .run](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/releases/download/cuda/NVIDIA-Linux-x86_64-545.23.08.run), ultimately fixing the Problem

> [!TIP]
> Another reason the Jailmaker alternative is recommended is because I had initially, when installing `steam-headless` with TrueCharts had kept my storage in PVC bindings, basically inaccessible, I then added mounts in an SMB share to access it there.

# Thank You
I would like to reiterate my thanks towards @alansari for helping me when it is easier not to and constantly helping within the issues that people face. Because of him I have bought a CCwGTV (ChromeCast with Google TV) and am really enjoying playing on Moonlight on the Big Screen.

**Comment by seisdr** [comment]
```
steam-headless-1  |   - Enable i386 arch
steam-headless-1  |   - Install mesa vulkan drivers
steam-headless-1  | **** Found NVIDIA device 'NVIDIA GeForce RTX 4070 Ti' ****
steam-headless-1  |   - Downloading driver v555.42.06
steam-headless-1 exited with code 0
```

**Comment by WillKirkmanM** [comment]
> ```
> steam-headless-1  |   - Enable i386 arch
> steam-headless-1  |   - Install mesa vulkan drivers
> steam-headless-1  | **** Found NVIDIA device 'NVIDIA GeForce RTX 4070 Ti' ****
> steam-headless-1  |   - Downloading driver v555.42.06
> steam-headless-1 exited with code 0
> ```

Hey @seisdr, in the TL;DR of my [solution post](https://github.com/Steam-Headless/docker-steam-headless/issues/152#issuecomment-2208788858), there is a repository by flathub where you can download the `.run` file for NVIDIA driver's that are not [listed in the repository](https://download.nvidia.com/XFree86/Linux-x86_64).

You can find that repository here: https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/releases/tag/cuda or on the first line of my solution. 
![image](https://github.com/Steam-Headless/docker-steam-headless/assets/98240335/2554da1a-8d01-42e8-adcc-a9e66ba9932e)

**Comment by seisdr** [comment]
It's not working docker compose is stuck in a loop
```
user@user:~/docker-steam-headless/docs/compose-files$ ls ~/Downloads/
NVIDIA_555.42.06.run
user@user:~/docker-steam-headless/docs/compose-files$ sudo docker compose up
WARN[0000] /home/user/docker-steam-headless/docs/compose-files/docker-compose.yml: `version` is obsolete
[+] Running 1/0
 ✔ Container compose-files-steam-headless-1  Created                                       0.0s
Attaching to steam-headless-1
steam-headless-1  | Build: [2024-06-29 02:44:04] [master] [9cf971f68f8be8b2f9490469e5b91ee8cf1e8e2d] [debian]
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/10-setup_user.sh: executing... ]
steam-headless-1  | **** Configure default user ****
steam-headless-1  |   - Setting default user uid=1000(default) gid=1000(default)
steam-headless-1  | usermod: no changes
steam-headless-1  |   - Adding default user to any additional required device groups
steam-headless-1  |   - Adding user 'default' to group: 'video'
steam-headless-1  |   - Adding user 'default' to group: 'audio'
steam-headless-1  |   - Adding user 'default' to group: 'input'
steam-headless-1  |   - Adding user 'default' to group: 'pulse'
steam-headless-1  |   - Setting umask to 000
steam-headless-1  |   - Create the user XDG_RUNTIME_DIR path '/tmp/.X11-unix/run'
steam-headless-1  |   - Setting ownership of all log files in '/home/default/.cache/log'
steam-headless-1  |   - Setting root password
steam-headless-1  |   - Setting user password
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/11-setup_sysctl_values.sh: executing... ]
steam-headless-1  | **** Configure some system kernel parameters ****
steam-headless-1  |   - The vm.max_map_count is already greater than '524288'
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_dbus.sh: executing... ]
steam-headless-1  | **** Configure container dbus ****
steam-headless-1  |   - Container configured to run its own dbus
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_udev.sh: executing... ]
steam-headless-1  | **** Configure udevd ****
steam-headless-1  |   - Disable udevd - /sys is mounted RO
steam-headless-1  |   - Enable dumb-udev service
steam-headless-1  |   - Ensure the default user has permission to r/w on input devices
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/40-setup_locale.sh: executing... ]
steam-headless-1  | **** Configure local ****
steam-headless-1  |   - Locales already set correctly to en_US.UTF-8 UTF-8
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/50-configure_pulseaudio.sh: executing... ]
steam-headless-1  | **** Configure pulseaudio ****
steam-headless-1  |   - Enable pulseaudio service.
steam-headless-1  |   - Configure pulseaudio to pipe audio to a socket
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/60-configure_gpu_driver.sh: executing... ]
steam-headless-1  | **** No Intel device found ****
steam-headless-1  | **** Found AMD device '0f:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Raphael (rev c9)' ****
steam-headless-1  |   - Mesa has already been installed into this container
steam-headless-1  | **** Found NVIDIA device 'NVIDIA GeForce RTX 4070 Ti' ****
steam-headless-1  |   - Downloading driver v555.42.06
steam-headless-1 exited with code 8
steam-headless-1  |   - Adding default user to any additional required device groups
steam-headless-1  |   - Adding user 'default' to group: 'video'
steam-headless-1  |   - Adding user 'default' to group: 'audio'
steam-headless-1  |   - Adding user 'default' to group: 'input'
steam-headless-1  |   - Adding user 'default' to group: 'pulse'
steam-headless-1  |   - Setting umask to 000
steam-headless-1  |   - Create the user XDG_RUNTIME_DIR path '/tmp/.X11-unix/run'
steam-headless-1  |   - Setting ownership of all log files in '/home/default/.cache/log'
steam-headless-1  |   - Setting root password
steam-headless-1  |   - Setting user password
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/11-setup_sysctl_values.sh: executing... ]
steam-headless-1  | **** Configure some system kernel parameters ****
steam-headless-1  |   - The vm.max_map_count is already greater than '524288'
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_dbus.sh: executing... ]
steam-headless-1  | **** Configure container dbus ****
steam-headless-1  |   - Container configured to run its own dbus
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_udev.sh: executing... ]
steam-headless-1  | **** Configure udevd ****
steam-headless-1  |   - Disable udevd - /sys is mounted RO
steam-headless-1  |   - Enable dumb-udev service
steam-headless-1  |   - Ensure the default user has permission to r/w on input devices
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/40-setup_locale.sh: executing... ]
steam-headless-1  | **** Configure local ****
steam-headless-1  |   - Locales already set correctly to en_US.UTF-8 UTF-8
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/50-configure_pulseaudio.sh: executing... ]
steam-headless-1  | **** Configure pulseaudio ****
steam-headless-1  |   - Enable pulseaudio service.
steam-headless-1  |   - Configure pulseaudio to pipe audio to a socket
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/60-configure_gpu_driver.sh: executing... ]
steam-headless-1  | **** No Intel device found ****
steam-headless-1  | **** Found AMD device '0f:00.0 VGA compatible controller: Advanced Micro Devices, Inc. [AMD/ATI] Raphael (rev c9)' ****
steam-headless-1  |   - Mesa has already been installed into this container
steam-headless-1  | **** Found NVIDIA device 'NVIDIA GeForce RTX 4070 Ti' ****
steam-headless-1  |   - Downloading driver v555.42.06
steam-headless-1  |   - Adding default user to any additional required device groups
steam-headless-1  |   - Adding user 'default' to group: 'video'
steam-headless-1  |   - Adding user 'default' to group: 'audio'
steam-headless-1  |   - Adding user 'default' to group: 'input'
steam-headless-1  |   - Adding user 'default' to group: 'pulse'
steam-headless-1  |   - Setting umask to 000
steam-headless-1  |   - Create the user XDG_RUNTIME_DIR path '/tmp/.X11-unix/run'
steam-headless-1  |   - Setting ownership of all log files in '/home/default/.cache/log'
steam-headless-1  |   - Setting root password
steam-headless-1  |   - Setting user password
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/11-setup_sysctl_values.sh: executing... ]
steam-headless-1  | **** Configure some system kernel parameters ****
steam-headless-1  |   - The vm.max_map_count is already greater than '524288'
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_dbus.sh: executing... ]
steam-headless-1  | **** Configure container dbus ****
steam-headless-1  |   - Container configured to run its own dbus
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/30-configure_udev.sh: executing... ]
steam-headless-1  | **** Configure udevd ****
steam-headless-1  |   - Disable udevd - /sys is mounted RO
steam-headless-1  |   - Enable dumb-udev service
steam-headless-1  |   - Ensure the default user has permission to r/w on input devices
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/40-setup_locale.sh: executing... ]
steam-headless-1  | **** Configure local ****
steam-headless-1  |   - Locales already set correctly to en_US.UTF-8 UTF-8
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/50-configure_pulseaudio.sh: executing... ]
steam-headless-1  | **** Configure pulseaudio ****
steam-headless-1  |   - Enable pulseaudio service.
steam-headless-1  |   - Configure pulseaudio to pipe audio to a socket
steam-headless-1  | DONE
steam-headless-1  |
steam-headless-1  | [ /etc/cont-init.d/60-configure_gpu_driver.sh: executing... ]
^CGracefully stopping... (press Ctrl+C again to force)
[+] Stopping 1/1
 ✔ Container compose-files-steam-headless-1  Stopped                                       0.6s
canceled
user@user:~/docker-steam-headless/docs/compose-files$
```

**Comment by WillKirkmanM** [comment]
I would recommend updating the NVIDIA driver's on your host system. Are you on TrueNas Scale?

**Comment by seisdr** [comment]
No just a Linux , if i install a higher driver i get a mismatch

**Comment by WillKirkmanM** [comment]
Try to install a higher driver, get that mismatch and then download the updated driver's .run and place it in the Downloads folder to fix the mismatch.

**Comment by WillKirkmanM** [comment]
This has been fixed in the merged pull request https://github.com/Steam-Headless/docker-steam-headless/pull/153.

**Comment by WillKirkmanM** [comment]
This issue has been fixed in the merged pull request https://github.com/Steam-Headless/docker-steam-headless/pull/153.
